File: ProjectTemplateSourceSolutions/VisualBasic/NoViewEngine/Containers/EndpointContainer.vb
Imports System.Linq.Expressions
Imports JuniorRouteWebApplication.Endpoints
Imports Junior.Route.AutoRouting.AntiCsrf.HtmlGenerators
Imports Junior.Route.Routing.AntiCsrf.HtmlGenerators
Imports Junior.Route.Routing.AntiCsrf.CookieManagers
Imports Junior.Route.Routing.AntiCsrf
Imports Junior.Route.AutoRouting
Imports Junior.Common
Imports Junior.Route.AutoRouting.Containers
Imports Junior.Route.Routing.AntiCsrf.NonceRepositories
Imports Junior.Route.Routing.AntiCsrf.NonceValidators
Imports Junior.Route.Routing.AntiCsrf.ResponseGenerators
Imports System.Reflection

Namespace Containers
	Public Class EndpointContainer
		Implements IContainer
		Private ReadOnly _concreteTypesByRequestedType As Dictionary(Of Type, Type) = New Dictionary(Of Type, Type)
		Private ReadOnly _instancesByRequestedType As Dictionary(Of Type, Object) = New Dictionary(Of Type, Object)
		Private ReadOnly _lockObject = New Object()

		Public Sub New()
			' Common
			AddMapping (Of IGuidFactory, GuidFactory)()
			AddMapping (Of ISystemClock, SystemClock)()
			AddMapping (Of IResponseContext, ResponseContext)()

			' Anti-CSRF
			AddMapping (Of IAntiCsrfConfiguration, ConfigurationSectionAntiCsrfConfiguration)()
			AddMapping (Of IAntiCsrfCookieManager, DefaultCookieManager)()
			AddMapping (Of IAntiCsrfHtmlGenerator, DefaultHtmlGenerator)()
			AddMapping (Of IAntiCsrfNonceRepository, MemoryCacheNonceRepository)()
			AddMapping (Of IAntiCsrfNonceValidator, DefaultNonceValidator)()
			AddMapping (Of IAntiCsrfResponseGenerator, DefaultResponseGenerator)()

			' Endpoints
			AddMapping (Of HelloWorld, HelloWorld)()
		End Sub

		Public Function GetInstance (Of T)() As T Implements IContainer.GetInstance
			Return GetInstance(GetType(T))
		End Function

		Public Function GetInstance(ByVal type As Type) As Object Implements IContainer.GetInstance
			type.ThrowIfNull("type")

			SyncLock _lockObject
				Dim existingInstance As Object = Nothing

				If _instancesByRequestedType.TryGetValue(type, existingInstance) Then
					Return existingInstance
				End If

				Dim concreteType As Type = Nothing

				If Not _concreteTypesByRequestedType.TryGetValue(type, concreteType) Then
					Return Nothing
				End If

				Dim constructorInfos As ConstructorInfo() = concreteType.GetConstructors(BindingFlags.Public + BindingFlags.Instance)

				If constructorInfos.Length = 0 Then
					Throw New ArgumentException(String.Format("No public instance constructors found for type {0}.", concreteType.FullName))
				End If

				Dim pair As KeyValuePair(Of ConstructorInfo, ParameterInfo()) = constructorInfos _
					    .Select(Function(arg) New With {.constructorInfo = arg, .parameterInfos = arg.GetParameters()}) _
					    .ToDictionary(Function(arg) arg.constructorInfo, Function(arg) arg.parameterInfos) _
					    .OrderByDescending(Function(arg) arg.Value.Length) _
					    .Single()
				Dim parameterValues As IEnumerable(Of Object) = pair.Value.Select(
					Function(parameterInfo)
						Dim parameterValue As Object = If(GetInstance(parameterInfo.ParameterType), If(parameterInfo.HasDefaultValue, parameterInfo.DefaultValue, Nothing))

						If (parameterValue Is Nothing) Then
							Throw New ApplicationException(String.Format("Unable to map parameter '{0}' of type {1}.", parameterInfo, pair.Key.DeclaringType.FullName))
						End If

						Return parameterValue
					End Function)
				Dim newExpression As NewExpression = Expression.[New](pair.Key, parameterValues.Select(AddressOf Expression.Constant))
				Dim newInstance = Expression.Lambda(newExpression).Compile().DynamicInvoke()

				_instancesByRequestedType.Add(type, newInstance)

				Return newInstance
			End SyncLock
		End Function

		Private Sub AddMapping (Of TRequestedType, TConcreteType As TRequestedType)()
			SyncLock _lockObject
				_concreteTypesByRequestedType.Add(GetType(TRequestedType), GetType(TConcreteType))
			End SyncLock
		End Sub
	End Class
End Namespace

File: ProjectTemplateSourceSolutions/CSharp/NoViewEngine/Containers/EndpointContainer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Reflection;

using Junior.Common;
using Junior.Route.AutoRouting;
using Junior.Route.AutoRouting.AntiCsrf.HtmlGenerators;
using Junior.Route.AutoRouting.Containers;
using Junior.Route.Routing.AntiCsrf;
using Junior.Route.Routing.AntiCsrf.CookieManagers;
using Junior.Route.Routing.AntiCsrf.HtmlGenerators;
using Junior.Route.Routing.AntiCsrf.NonceRepositories;
using Junior.Route.Routing.AntiCsrf.NonceValidators;
using Junior.Route.Routing.AntiCsrf.ResponseGenerators;

using JuniorRouteWebApplication.Endpoints;

namespace JuniorRouteWebApplication.Containers
{
	// This container does not support concrete types with generic type arguments
	// Developers are advised to replace this container with a more robust container, if necessary
	public class EndpointContainer : IContainer
	{
		private readonly Dictionary<Type, Type> _concreteTypesByRequestedType = new Dictionary<Type, Type>();
		private readonly Dictionary<Type, object> _instancesByRequestedType = new Dictionary<Type, object>();
		private readonly object _lockObject = new object();

		public EndpointContainer()
		{
			// Common
			AddMapping<IGuidFactory, GuidFactory>();
			AddMapping<ISystemClock, SystemClock>();
			AddMapping<IResponseContext, ResponseContext>();

			// Anti-CSRF
			AddMapping<IAntiCsrfConfiguration, ConfigurationSectionAntiCsrfConfiguration>();
			AddMapping<IAntiCsrfCookieManager, DefaultCookieManager>();
			AddMapping<IAntiCsrfHtmlGenerator, DefaultHtmlGenerator>();
			AddMapping<IAntiCsrfNonceRepository, MemoryCacheNonceRepository>();
			AddMapping<IAntiCsrfNonceValidator, DefaultNonceValidator>();
			AddMapping<IAntiCsrfResponseGenerator, DefaultResponseGenerator>();

			// Endpoints
			AddMapping<HelloWorld, HelloWorld>();
		}

		public T GetInstance<T>()
		{
			return (T)GetInstance(typeof(T));
		}

		public object GetInstance(Type type)
		{
			type.ThrowIfNull("type");

			lock (_lockObject)
			{
				object existingInstance;

				if (_instancesByRequestedType.TryGetValue(type, out existingInstance))
				{
					return existingInstance;
				}

				Type concreteType;

				if (!_concreteTypesByRequestedType.TryGetValue(type, out concreteType))
				{
					return null;
				}

				ConstructorInfo[] constructorInfos = concreteType.GetConstructors(BindingFlags.Public | BindingFlags.Instance);

				if (constructorInfos.Length == 0)
				{
					throw new ArgumentException(String.Format("No public instance constructors found for type {0}.", concreteType.FullName));
				}

				KeyValuePair<ConstructorInfo, ParameterInfo[]> pair = constructorInfos
					.Select(arg => new { constructorInfo = arg, parameterInfos = arg.GetParameters() })
					.ToDictionary(arg => arg.constructorInfo, arg => arg.parameterInfos)
					.OrderByDescending(arg => arg.Value.Length)
					.Single();
				IEnumerable<object> parameterValues = pair.Value.Select(parameterInfo =>
					{
						object parameterValue = GetInstance(parameterInfo.ParameterType) ?? (parameterInfo.HasDefaultValue ? parameterInfo.DefaultValue : null);

						if (parameterValue == null)
						{
							throw new ApplicationException(String.Format("Unable to map parameter '{0}' of type {1}.", parameterInfo, pair.Key.DeclaringType.FullName));
						}

						return parameterValue;
					});
				NewExpression newExpression = Expression.New(pair.Key, parameterValues.Select(Expression.Constant));
				object newInstance = Expression.Lambda(newExpression).Compile().DynamicInvoke();

				_instancesByRequestedType.Add(type, newInstance);

				return newInstance;
			}
		}

		private void AddMapping<TRequestedType, TConcreteType>()
			where TConcreteType : TRequestedType
		{
			lock (_lockObject)
			{
				_concreteTypesByRequestedType.Add(typeof(TRequestedType), typeof(TConcreteType));
			}
		}
	}
}